ITR/data/base_providers.py

```diff
@@ -107,6 +111,12 @@ def __init__(self, EI_benchmarks: IEIBenchmarkScopes,
         self._EI_benchmarks = EI_benchmarks
         self.temp_config = tempscore_config
         self.column_config = column_config
+        # Benchmark's primary scope, for which we calculate
```

Comment:
I think there's a misunderstanding here.  A benchmark either has S1+S2 as a two-layer sandwich or S1+S2+S3 as a three-layer sandwich.  There is an asterisk on the menu: not all scope 3 disclosures possible are necessarily covered by the benchmark.  If the benchmark says "Scope 3 is defined as exactly the Scope 2 emissions of buildings" then that's that, and we don't need or want to add additional Scope3 elements to the mix.

But in any event, just because one benchmark has S3 emissions counted in a separate bucket and another folds them into S1 doesn't mean that there's primacy of one over another.   I'll continue reviewing and see if I come back to this point.

Reply:
I see, I think my use of the term "primary" is confusing and not correct.
What I mean is: our tool shows only one scope. The loaded intensity benchmark defines, what it will be. When default benchmarks are loaded - we need to calculate S1S2, and when the "S3 separate" benchmark is loaded - need to calculate S3.
I used the term "primary", to describe, which scope we need to calculate for a loaded benchmark file.
I will rename it to something less confusing and more neutral

Reply:
It is a feature of the tool to support the analysis under multiple benchmarks, producing temperature scores for portfolios labeled with the benchmarks used as input.  That said, we have production (not scope dependent), emissions (completely scope dependent) and emissions intensity.  A company decides whether to "build or buy" and in so doing, may decide whether to be an emitter (S1, build) or user (S3, buy).  A sectoral emission intensity will bake in averages of buy vs. build, which needs to be unbaked to align with whatever the company is actually reporting.

But I now see a second potential problem (coming from the original code).  The functions that deliver a company's projected EIs (whether targets or trajectories) must not fold everything into S1S2, but keep the scopes in whatever way they were.  Some parts of the code in base_providers.py made this data wide (with S1S2 and S3 part of the same row, see get_company_intensity_and_production_at_base_year), but the logic of the temperature score calculations is more based on long data (expecting there to be a single intensity, see get_company_projected_trajectories and get_company_projected_targets).  Those wrong trajectories are going to mess up the calculations of overshoot ratios and ultimately temperature score.

So you need to trace that you can feed a company with separate S1, S2, and S3 emissions through to individual temperature scores that charge against individual benchmark budgets than can then be emissions-weighted.  Why emissions-weighted?  Because the ITR temp score computes its score based only on budgetary allocations.  If a company overshoots S1 by 10%, undershoots S2 by 20% and overshoots S3 by 30%, you'll get three different temperature scores that need to be ultimately combined into one.  If S3 emissions are 90% of the total emissions, you don't want to simply average them!  The score has to make sense not only in the individual scope context (which might be S1S2 or S1S2S3 if that's the input and the benchmark), but also as a composite result when appropriate.

Reply:
> It is a feature of the tool to support the analysis under multiple benchmarks

> So you need to trace that you can feed a company with separate S1, S2, and S3 emissions through to individual temperature scores that charge against individual benchmark budgets

It is a target feature, yes, but it is not implemented. The implementation supports a single scope S1S2.

We may see 'primary_scope' as a temporal solution, to unblock progress for "OECM S3 separate" calculation. The full feature support we already called a "scopes epic", and we will approach it eventually. But we identified higher priority for S3 separate calculation, and you added a separate JSON file, to support this path.

Let's proceed with "primary scope", to achieve helpful results short-term, and support full scopes support under continuous development paradigm long-term.

> can then be emissions-weighted. Why emissions-weighted? Because the ITR temp score computes its score based only on budgetary allocations. If a company overshoots S1 by 10%, undershoots S2 by 20% and overshoots S3 by 30%, you'll get three different temperature scores that need to be ultimately combined into one. If S3 emissions are 90% of the total emissions, you don't want to simply average them! 

I have hard times formulating this logic as an algorithm. Could you point me to the pdf, which would describe the algorithm, as defined by the methodology?

Reply:
Here's a diagram: https://github.com/os-climate/ITR/blob/main/docs/Calculation.md

Here is [the TCFD report, in pdf format](https://www.tcfdhub.org/wp-content/uploads/2021/10/PAT_Measuring_Portfolio_Alignment_Technical_Considerations.pdf) from which all ITR tools take their guidance.

Reply:
> Here's a diagram: https://github.com/os-climate/ITR/blob/main/docs/Calculation.md

it only shows the need of weighted values, but doesn't define, which values the weights have

> Here is [the TCFD report, in pdf format](https://www.tcfdhub.org/wp-content/uploads/2021/10/PAT_Measuring_Portfolio_Alignment_Technical_Considerations.pdf) from which all ITR tools take their guidance.

i see it mentions, that inputs need to be weighted, but i don't see, where it specifies the weights values.

> I have hard times formulating this logic as an algorithm. Could you point me to the pdf, which would describe the algorithm, as defined by the methodology?

this statement is still valid: i don't know, which weights to apply. You pointed me to the concepts; could you point me to *algorithms*?

Reply:
TCRE is explained here https://en.wikipedia.org/wiki/Transient_climate_response_to_cumulative_carbon_emissions and underlying carbon budgets here: https://en.wikipedia.org/wiki/Carbon_budget

Translating to an algorithm, if we use the 500 Gt carbon budget for 1.5 deltaDegC target and we are already at 1.2C deltaDegC we have 0.3 deltaDegC of warming left.   0.3C / 500 Gt = 0.0006 deltaDegC / Gt CO2e
The benchmarks attempt to allocate the 500 Gt of "remaining carbon budget" as fairly as possible to allow a soft landing of the economy by 2050.  Such a landing requires us to cut CO2e emissions 50% by 2030...just saying.

In terms of an algorithm, everything is about answering the question "if all companies did what this company was doing, projected through 2050, what would be temperature rise based on TCRE?"  So the algorithm needs to calculate the given companies emissions and compare to the budget, which is given sector by sector.  Some sectors don't emit much, and they need to keep not emitting much.

The total emissions TE = sum(scope emissions), or sum(S1E, S2E, S3E).
Scope emissions are a function of production: Scope E = P * Scope EI
Thus TE = P*S1EI + P*S2EI * P*S3EI = P*(S1EI + S2EI + S3EI).  Alas, EI is not a constant number, but depends on scope, thus S1EI is not scope 1 emissions times a constant EI, but is its own metric.

Let EB be total Emissions Budget, and S1EB, S2EB, and S3EB be the respective scope emission budgets
TS (Temperature score) = TCRE * (TE - EB)

I believe TS can be decomposed: TS(TE) =
    TS(S1E - S1E Budget)*(S1EB/EB) + TS(S2E - S2EB)*(S2EB/EB) + TS(S3E - S3EB)*(S3EB/EB)

You can see that is a weighted average, and I believe (tests will reveal) that the equality holds.  Does that answer the algorithm question?

Reply:
i pushed an easy part of this comment so far: renamed "primary_scope" into "scope_to_calc".
i plan to figure out, how to do the weights, later this week

Reply:
@MichaelTiemannOSC all right, Michael, here is a catch:
The class method `TemperatureScore._get_aggregations()` uses local variables, called `weighted_scores`.
These local variables internally call `PortfolioAggregation._calculate_aggregate_score()`, which take in account S1S2 and S3.

Doesn't it mean, that the weights already apply?

Reply:
PortfolioAggregation._calculate_aggregate_score calculates *portfolio* weighting, which can be based on company valuation, investment size, company revenues, total emissions, etc.  It's a different kind of weight.

I also don't like what I see at lines 114-115:

```
                return ((s1s2[self.c.COLS.TEMPERATURE_SCORE] * s1s2[self.c.COLS.GHG_SCOPE12] +
                         s3[self.c.COLS.TEMPERATURE_SCORE] * s3[self.c.COLS.GHG_SCOPE3]) / company_emissions)
```

should be

```
                return ((s1s2[self.c.COLS.TEMPERATURE_SCORE] * CUMULATIVE(s1s2) +
                         s3[self.c.COLS.TEMPERATURE_SCORE] * CUMULATIVE(s3) / company_emissions)
```

Where CUMULATIVE calculates the weighting between self.column_config.CUMULATIVE_TRAJECTORY and self.column_config.CUMULATIVE_TARGET based on TARGET_PROBABILITY (which by default is 50/50).

The reason the current code is wrong is because it weights based on the company's current s1s2/s3 weighting, but the temperature score is based on cumulative emissions, so we need to get the cumulative weights of those two.

Reply:
> It's a different kind of weight.

i pushed a new commit, which adds weights to temperature score

> I also don't like what I see at lines 114-115

i'd rather not touch them in this PR. Let's create a GitHub "issue", and resolve it as a separate task

Reply:
This comment thread hold a few different discussion points.
Some of them we resolved within this thread.
Some of them we discussed in slack and calls, and resolved.
Some of them we duplicate in other threads within this PR, where we continue the discussions.

For these 3 reasons, and to keep the precise more clean, I close this thread as resolved. Please let me know, if you think we should re-open it